hazelcast/src/main/java/com/hazelcast/query/Predicates.java

```diff
@@ -38,68 +38,215 @@
     private Predicates() {
     }
 
+    /**
+     * Creates a new instance of <b>instance of</b> predicate.
+     * <p>
+     * The created predicate behaves exactly the same as the Java {@code instanceof} operator.
+     *
+     * @param klass the class the created predicate will check for.
+     * @return the created <b>instance of</b> predicate instance.
+     */
     public static Predicate instanceOf(final Class klass) {
         return new InstanceOfPredicate(klass);
     }
 
+    /**
+     * Creates a new instance of <b>and</b> predicate.
+     * <p>
+     * The created predicate performs the logical <b>and</b> operation on the given {@code predicates}.
+     *
```

Comment:
Add something like this:
>With zero {@code predicates} it always passes. With single predicate it passes, when the given predicate passes. I'm not sure if the behavior is the same with index.
>Not sure about short-circuit evaluation: if the first expression in `false`, second is not evaluated at all. I couldn't find an example when it will make difference.

Similar for `or` (except it returns false with zero predicates).

Reply:
Good point. The only question, is the short-circuiting behavior is considered to be a part of the public contract or not. If we document it, we will have to support it. WDYT?

Reply:
I would avoid adding implementation details such as short-circuit evaluation and possibly even usage of the index. These details can depend on a lot of things such as usage of HD memory, dependance on method parameters and some inadvertent assumptions in the implementation that can later change. Adding them makes the javadoc long and brittle as it is guaranteed to go stale much faster. The javadoc should be short and list the invariants. In this sense, it makes sense to add information such as handling `null`, type conversion, escaping and special keywords such as `__key` and `this`.

Reply:
Short-circuiting is probably not needed, as the predicates never fail, so id doesn't matter if the second is executed or not.

But the index usage should definitely be documented. Indexes should speed up the searches from seconds to milliseconds (that is 1000x), you are definitely interested of what you can and cannot do. Seems to me that indexes isn't a feature used much, because it's very under-documented.

Reply:
But OK, it's maybe out of scope for this basic documentation endeavor. There might be dragoons :)

Reply:
Added three sections to Predicates class javadoc: Special Attributes, Handling of null and Type Conversion.